Unfortunately this causes a linker error with [Higgs](https://github.com/higgsjs/Higgs)' test.
```
higgs-test.o:ffi/tests.d:function _D3std5range14__T6strideTAkZ6strideFAkmZ6Result7popBackMFNaNbNiNfZv: error: undefined reference to '_D3std5range10primitives16__T8popBackNTAkZ8popBackNFNaNbNiNfKAkmZm'
higgs-test.o:ffi/tests.d:function _D3std5range14__T6strideTAkZ6strideFAkmZ6Result6moveAtMFNaNbNiNfmZk: error: undefined reference to '_D3std5range10primitives16__T6moveAtTAkTmZ6moveAtFNaNbNiNfAkmZk'
```

I'll try to provide more details, but it seems that some phobos unittest (std.uni) gets compiled into the binary but some templates instantiated by it are not. Might be the `-debug`, `-unittest` issues mentioned in needsCodegen.

--- a/src/aggregate.h
+++ b/src/aggregate.h
@@ -170,6 +170,11 @@ public:
     Type *arg1type;
     Type *arg2type;
 
+    // Even if struct is defined as non-root symbol, some built-in operations
+    // (e.g. TypeidExp, NewExp, ArrayLiteralExp, etc) request its TypeInfo.
+    // For those, today TypeInfo_Struct is generated in COMDAT.
+    bool requestTypeInfo;
+
     StructDeclaration(Loc loc, Identifier *id);
     Dsymbol *syntaxCopy(Dsymbol *s);
     void semantic(Scope *sc);

--- a/src/expression.c
+++ b/src/expression.c
@@ -4161,7 +4161,7 @@ Expression *ArrayLiteralExp::semantic(Scope *sc)
         return new ErrorExp();
     }
 
-    semanticTypeInfo(sc, t0);
+    semanticTypeInfo(sc, type);
 
     return this;
 }
@@ -6114,6 +6114,9 @@ Expression *TypeidExp::semantic(Scope *sc)
         // Handle this in the glue layer
         e = new TypeidExp(loc, ta);
         e->type = getTypeInfoType(ta, sc);
+
+        semanticTypeInfo(sc, ta);
+
         if (ea)
         {
             e = new CommaExp(loc, ea, e);       // execute ea

--- a/src/magicport.json
+++ b/src/magicport.json
@@ -248,6 +248,7 @@
                 "aggregate",
                 "argtypes",
                 "arraytypes",
+                "backend",
                 "clone",
                 "declaration",
                 "dmodule",

--- a/src/struct.c
+++ b/src/struct.c
@@ -24,6 +24,7 @@
 #include "template.h"
 #include "tokens.h"
 
+Type *getTypeInfoType(Type *t, Scope *sc);
 TypeTuple *toArgTypes(Type *t);
 
 FuncDeclaration *StructDeclaration::xerreq;     // object.xopEquals
@@ -105,11 +106,19 @@ void semanticTypeInfo(Scope *sc, Type *t)
                 if (ti->minst && !ti->minst->isRoot())
                     Module::addDeferredSemantic3(sd);
             }
-            else if (sd->inNonRoot())
+            else
             {
-                //printf("deferred sem3 for TypeInfo - sd = %s, inNonRoot = %d\n", sd->toChars(), sd->inNonRoot());
-                Module::addDeferredSemantic3(sd);
+                if (sd->inNonRoot())
+                {
+                    //printf("deferred sem3 for TypeInfo - sd = %s, inNonRoot = %d\n", sd->toChars(), sd->inNonRoot());
+                    Module::addDeferredSemantic3(sd);
+                }
             }
+
+            getTypeInfoType(t, sc);
+
+            if (sc->minst)
+                sd->requestTypeInfo = true;
         }
         void visit(TypeClass *t) { }
         void visit(TypeTuple *t)
@@ -667,6 +676,7 @@ StructDeclaration::StructDeclaration(Loc loc, Identifier *id)
     ispod = ISPODfwd;
     arg1type = NULL;
     arg2type = NULL;
+    requestTypeInfo = false;
 
     // For forward references
     type = new TypeStruct(this);

--- a/src/template.c
+++ b/src/template.c
@@ -7263,6 +7263,14 @@ Dsymbols *TemplateInstance::appendToModuleMember()
 {
     Module *mi = minst;     // instantiated -> inserted module
 
+    if (global.params.useUnitTests ||
+        global.params.debuglevel)
+    {
+        // Turn all non-root instances to speculative
+        if (mi && !mi->isRoot())
+            mi = NULL;
+    }
+
     if (!mi || mi->isRoot())
     {
         /* If the instantiated module is speculative or root, insert to the
@@ -7855,15 +7863,8 @@ void unSpeculative(Scope *sc, RootObject *o)
  */
 bool TemplateInstance::needsCodegen()
 {
-    /* The issue is that if the importee is compiled with a different -debug
-     * setting than the importer, the importer may believe it exists
-     * in the compiled importee when it does not, when the instantiation
-     * is behind a conditional debug declaration.
-     */
-    // workaround for Bugzilla 11239
-    if (global.params.useUnitTests ||
-        global.params.allInst ||
-        global.params.debuglevel)
+    // Now -allInst is just for the backward compatibility.
+    if (global.params.allInst)
     {
         //printf("%s minst = %s, enclosing (%s)->isNonRoot = %d\n",
         //    toPrettyChars(), minst ? minst->toChars() : NULL,
@@ -7894,9 +7895,12 @@ bool TemplateInstance::needsCodegen()
         return true;
     }
 
-    // If this may be a speculative instantiation:
     if (!minst)
     {
+        // If this is a speculative instantiation,
+        // 1. do codegen if ancestors really needs codegen.
+        // 2. become non-speculative if siblings are not speculative
+
         TemplateInstance *tnext = this->tnext;
         TemplateInstance *tinst = this->tinst;
         // At first, disconnect chain first to prevent infinite recursion.
@@ -7911,6 +7915,42 @@ bool TemplateInstance::needsCodegen()
             assert(minst->isRoot() || minst->rootImports());
             return true;
         }
+        if (tnext && (tnext->needsCodegen() || tnext->minst))
+        {
+            minst = tnext->minst;   // cache result
+            assert(minst);
+            return minst->isRoot() || minst->rootImports();
+        }
+
+        // Elide codegen because this is really speculative.
+        return false;
+    }
+
+    /* The issue is that if the importee is compiled with a different -debug
+     * setting than the importer, the importer may believe it exists
+     * in the compiled importee when it does not, when the instantiation
+     * is behind a conditional debug declaration.
+     */
+    // workaround for Bugzilla 11239
+    if (global.params.useUnitTests ||
+        global.params.debuglevel)
+    {
+        // Prefer instantiations from root modules, to maximize link-ability.
+        if (minst->isRoot())
+            return true;
+
+        TemplateInstance *tnext = this->tnext;
+        TemplateInstance *tinst = this->tinst;
+        this->tnext = NULL;
+        this->tinst = NULL;
+
+        if (tinst && tinst->needsCodegen())
+        {
+            minst = tinst->minst;   // cache result
+            assert(minst);
+            assert(minst->isRoot() || minst->rootImports());
+            return true;
+        }
         if (tnext && tnext->needsCodegen())
         {
             minst = tnext->minst;   // cache result
@@ -7918,12 +7958,33 @@ bool TemplateInstance::needsCodegen()
             assert(minst->isRoot() || minst->rootImports());
             return true;
         }
+
+        // Bugzilla 2500 case
+        if (minst->rootImports())
+            return true;
+
+        // Elide codegen because this is not included in root instances.
         return false;
     }
-
-    if (minst->isRoot())
+    else
     {
-        // Prefer instantiation in non-root module, to minimize object code size
+        // Prefer instantiations from non-root module, to minimize object code size.
+
+        /* If a TemplateInstance is ever instantiated by non-root modules,
+         * we do not have to generate code for it,
+         * because it will be generated when the non-root module is compiled.
+         *
+         * But, if the non-root 'minst' imports any root modules, it might still need codegen.
+         *
+         * The problem is if A imports B, and B imports A, and both A
+         * and B instantiate the same template, does the compilation of A
+         * or the compilation of B do the actual instantiation?
+         *
+         * See Bugzilla 2500.
+         */
+        if (!minst->isRoot() && !minst->rootImports())
+            return false;
+
         TemplateInstance *tnext = this->tnext;
         this->tnext = NULL;
 
@@ -7933,29 +7994,10 @@ bool TemplateInstance::needsCodegen()
             assert(!minst->isRoot());
             return false;
         }
-    }
-    else
-    {
-        /* If a TemplateInstance is ever instantiated by non-root modules,
-         * we do not have to generate code for it,
-         * because it will be generated when the non-root module is compiled.
-         *
-         * But, if minst imports any root modules, we still need to generate the code.
-         *
-         * The problem is if A imports B, and B imports A, and both A
-         * and B instantiate the same template, does the compilation of A
-         * or the compilation of B do the actual instantiation?
-         *
-         * See bugzilla 2500.
-         */
 
-        if (!minst->rootImports())
-        {
-            //printf("instantiated by %s   %s\n", minst->toChars(), toChars());
-            return false;
-        }
+        // Do codegen because this is not included in non-root instances.
+        return true;
     }
-    return true;
 }
 
 /* ======================== TemplateMixin ================================ */

--- a/src/template.c
+++ b/src/template.c
@@ -5942,6 +5942,17 @@ Lerror:
         this->tnext = inst->tnext;
         inst->tnext = this;
 
+        /* A module can have explicit template instance and its alias
+         * in module scope (e,g, `alias Base64 = Base64Impl!('+', '/');`).
+         * If the first instantiation 'inst' had happened in non-root module,
+         * compiler can assume that its instantiated code would be included
+         * in the separately compiled obj/lib file (e.g. phobos.lib).
+         *
+         * However, if 'this' second instantiation happened in root module,
+         * compiler might need to invoke its codegen (Bugzilla 2500 & 2644).
+         * But whole import graph is not determined until all semantic pass finished,
+         * so 'inst' should conservatively finish the semantic3 pass for the codegen.
+         */
         if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
         {
             /* Swap the position of 'inst' and 'this' in the instantiation graph.
@@ -7246,61 +7257,34 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
 }
 
 /*****************************************
- * Append 'this' to the enclosing module members[] so the semantic routines
- * will get called on the instance members.
+ * Append 'this' to the specific module members[]
  */
 Dsymbols *TemplateInstance::appendToModuleMember()
 {
-    Module *md = tempdecl->getModule();
+    Module *mi = minst;     // instantiated -> inserted module
 
-    Module *mi = minst;
-    if (mi)
+    if (!mi || mi->isRoot())
     {
-        if (mi->isRoot())
-        {
-            /* If both modules mi and md where the template is declared are
-             * roots, instead use md to store the generated code in it.
-             */
-            if (md && md->isRoot())
-                mi = md;
-        }
-        else
-        {
-            /* A module can have explicit template instance and its alias
-             * in module scope (e,g, `alias Base64Impl!('+', '/') Base64;`).
-             * When the module is just imported, normally compiler can assume that
-             * its instantiated code would be contained in the separately compiled
-             * obj/lib file (e.g. phobos.lib).
-             */
-            /*
-             * Bugzilla 2644: However, if the template is instantiated in both
-             * modules of root and non-root, compiler should generate its objcode.
-             * Therefore, always conservatively insert this instance to the member of
-             * a root module, then calculate the necessity by TemplateInstance::needsCodegen().
-             */
-        }
-        //printf("\t1: adding to %s\n", mi->toPrettyChars());
+        /* If the instantiated module is speculative or root, insert to the
+         * member of a root module. Then:
+         *  - semantic3 pass will get called on the instance members.
+         *  - codegen pass will get a selection chance to do/skip it.
+         */
+
+        // insert target is made stable by using the module
+        // where tempdecl is declared.
+        mi = tempdecl->getModule();
+        if (!mi->isRoot())
+            mi = mi->importedFrom;
+        assert(mi->isRoot());
     }
     else
     {
-        // 'this' is speculative instance
-        if (md && md->isRoot())
-        {
-            mi = md;
-        }
-        else
-        {
-            // select arbitrary root module
-            Dsymbol *s = enclosing ? enclosing : tempdecl->parent;
-            while (s && !s->isModule())
-                s = s->toParent2();
-            assert(s);
-            mi = (Module *)s;
-            if (!mi->isRoot())
-                mi = mi->importedFrom;
-        }
-        assert(mi->isRoot());
-        //printf("\t2: adding to module %s instead of module %s\n", mi->toPrettyChars(), sc->module->toPrettyChars());
+        /* If the instantiated module is non-root, insert to the member of the
+         * non-root module. Then:
+         *  - semantic3 pass won't be called on the instance.
+         *  - codegen pass won't reach to the instance.
+         */
     }
 
     Dsymbols *a = mi->members;

--- a/src/toobj.c
+++ b/src/toobj.c
@@ -58,6 +58,7 @@ Symbol *toVtblSymbol(ClassDeclaration *cd);
 Symbol *toInitializer(AggregateDeclaration *ad);
 Symbol *toInitializer(EnumDeclaration *ed);
 void genTypeInfo(Type *t, Scope *sc);
+bool isSpeculativeType(Type *t);
 
 void toDebug(EnumDeclaration *ed);
 void toDebug(StructDeclaration *sd);
@@ -1010,6 +1011,11 @@ void toObjFile(Dsymbol *ds, bool multiobj)
 
         void visit(TypeInfoDeclaration *tid)
         {
+            if (isSpeculativeType(tid->tinfo))
+            {
+                //printf("-speculative '%s'\n", tid->toPrettyChars());
+                return;
+            }
             //printf("TypeInfoDeclaration::toObjFile(%p '%s') protection %d\n", tid, tid->toChars(), tid->protection);
 
             if (multiobj)

--- a/src/typinf.c
+++ b/src/typinf.c
@@ -78,19 +78,9 @@ void genTypeInfo(Type *torig, Scope *sc)
             // Generate COMDAT
             if (sc)                     // if in semantic() pass
             {
-                if (sc->func && sc->func->inNonRoot())
-                {
-                    // Bugzilla 13043: Avoid linking TypeInfo if it's not
-                    // necessary for root module compilation
-                }
-                else
-                {
-                    // Find module that will go all the way to an object file
-                    Module *m = sc->module->importedFrom;
-                    m->members->push(t->vtinfo);
-
-                    semanticTypeInfo(sc, t);
-                }
+                // Find module that will go all the way to an object file
+                Module *m = sc->module->importedFrom;
+                m->members->push(t->vtinfo);
             }
             else                        // if in obj generation pass
             {
@@ -133,6 +123,96 @@ TypeInfoDeclaration *getTypeInfoDeclaration(Type *t)
     default:
         return TypeInfoDeclaration::create(t, 0);
     }
+}
+
+bool isSpeculativeType(Type *t)
+{
+    class SpeculativeTypeVisitor : public Visitor
+    {
+    public:
+        StructDeclaration *result;
+
+        SpeculativeTypeVisitor() : result(NULL) {}
+
+        void visit(Type *t)
+        {
+            Type *tb = t->toBasetype();
+            if (tb != t)
+                tb->accept(this);
+        }
+        void visit(TypeNext *t)
+        {
+            if (t->next)
+                t->next->accept(this);
+        }
+        void visit(TypeBasic *t) { }
+        void visit(TypeVector *t)
+        {
+            t->basetype->accept(this);
+        }
+        void visit(TypeAArray *t)
+        {
+            t->index->accept(this);
+            visit((TypeNext *)t);
+        }
+        void visit(TypeFunction *t)
+        {
+            visit((TypeNext *)t);
+            // Currently TypeInfo_Function doesn't store parameter types.
+        }
+        void visit(TypeStruct *t)
+        {
+            result = t->sym;
+        }
+        void visit(TypeClass *t) { }
+        void visit(TypeTuple *t)
+        {
+            if (t->arguments)
+            {
+                for (size_t i = 0; i < t->arguments->dim; i++)
+                {
+                    Type *tprm = (*t->arguments)[i]->type;
+                    if (tprm)
+                        tprm->accept(this);
+                }
+            }
+        }
+    };
+    SpeculativeTypeVisitor v;
+    t->accept(&v);
+
+    StructDeclaration *sd = v.result;
+    if (!sd)
+        return false;
+
+    if (TemplateInstance *ti = sd->isInstantiated())
+    {
+        if (!ti->needsCodegen())
+        {
+            /* Bugzilla 14425: TypeInfo_Struct would refer the members of
+             * struct (e.g. opEquals via xopEquals field), so if it's instantiated
+             * in speculative context, TypeInfo creation should also be
+             * stopped to avoid 'unresolved symbol' linker errors.
+             */
+            if (!ti->minst)
+                return true;
+
+            /* When -debug/-unittest is specified, all of non-root instances are
+             * automatically changed to speculative, and here is always reached
+             * from those instantiated non-root structs.
+             * Therefore, if the TypeInfo is not auctually requested,
+             * we have to elide its codegen.
+             */
+            if (!sd->requestTypeInfo)
+                return true;
+        }
+    }
+    else
+    {
+        //assert(!sd->inNonRoot() || sd->requestTypeInfo);  // valid?
+    }
+
+    return false;
 }
 
 /****************************************************
@@ -422,13 +502,27 @@ public:
 
         if (TemplateInstance *ti = sd->isInstantiated())
         {
-            /* Bugzilla 14425: TypeInfo_Struct would refer the members of
-             * struct (e.g. opEquals via xopEquals field), so if it's instantiated
-             * in speculative context, TypeInfo creation should also be
-             * stopped to avoid 'unresolved symbol' linker errors.
-             */
-            if (!ti->needsCodegen() && !ti->minst)
-                return;
+            if (!ti->needsCodegen())
+            {
+                assert(ti->minst);
+                assert(sd->requestTypeInfo);
+
+                /* ti->toObjFile() won't get called. So, store these
+                 * member functions into object file in here.
+                 */
+                if (sd->xeq && sd->xeq != StructDeclaration::xerreq)
+                    toObjFile(sd->xeq, global.params.multiobj);
+                if (sd->xcmp && sd->xcmp != StructDeclaration::xerrcmp)
+                    toObjFile(sd->xcmp, global.params.multiobj);
+                if (FuncDeclaration *ftostr = search_toString(sd))
+                    toObjFile(ftostr, global.params.multiobj);
+                if (sd->xhash)
+                    toObjFile(sd->xhash, global.params.multiobj);
+                if (sd->postblit)
+                    toObjFile(sd->postblit, global.params.multiobj);
+                if (sd->dtor)
+                    toObjFile(sd->dtor, global.params.multiobj);
+            }
         }
 
         /* Put out:

--- a/test/runnable/extra-files/linkdebug.d
+++ b/test/runnable/extra-files/linkdebug.d
@@ -1,0 +1,16 @@
+module linkdebug;
+
+void main()
+{
+    import linkdebug_uni;
+    import linkdebug_range;
+
+    // OK
+    //SortedRangeX!(uint[], "a <= b") SR;
+
+    CodepointSet set;
+    set.addInterval(1, 2);
+
+    // NG, order dependent.
+    SortedRange!(uint[], "a <= b") SR;
+}

--- a/test/runnable/extra-files/linkdebug_primitives.d
+++ b/test/runnable/extra-files/linkdebug_primitives.d
@@ -1,0 +1,13 @@
+module linkdebug_primitives;
+
+size_t popBackN(R)(ref R r, size_t n)
+{
+    n = cast(size_t) (n < r.length ? n : r.length);
+    r = r[0 .. $ - n];
+    return n;
+}
+
+auto moveAt(R, I)(R r, I i)
+{
+    return r[i];
+}

--- a/test/runnable/extra-files/linkdebug_range.d
+++ b/test/runnable/extra-files/linkdebug_range.d
@@ -1,0 +1,48 @@
+module linkdebug_range;
+
+import linkdebug_primitives : popBackN, moveAt;
+
+auto stride(R)(R r)
+{
+    static struct Result
+    {
+        R source;
+
+        void popBack()
+        {
+            popBackN(source, 0);
+        }
+
+        uint moveAt(size_t n)
+        {
+            return .moveAt(source, n);
+        }
+    }
+    return Result(r);
+}
+
+struct SortedRange(Range, alias pred = "a < b")
+{
+    this(Range input)
+    out
+    {
+        dbgVerifySorted();
+    }
+    body
+    {
+    }
+
+    void dbgVerifySorted()
+    {
+        debug
+        {
+            uint[] _input;
+            auto st = stride(_input);
+        }
+    }
+}
+
+auto assumeSorted(alias pred = "a < b", R)(R r)
+{
+    return SortedRange!(R, pred)(r);
+}

--- a/test/runnable/extra-files/linkdebug_uni.d
+++ b/test/runnable/extra-files/linkdebug_uni.d
@@ -1,0 +1,19 @@
+module linkdebug_uni;
+
+import linkdebug_range;
+
+struct GcPolicy {}
+
+alias CodepointSet = InversionList!();
+
+struct InversionList(SP = GcPolicy)
+{
+@trusted:
+    size_t addInterval(int a, int b, size_t hint = 0)
+    {
+        auto data = new uint[](0);        // affects to the number of missimg symbol
+        auto range = assumeSorted(data[]);  // NG
+        //SortedRange!(uint[], "a < b") SR; // OK
+        return 1;
+    }
+}

--- a/test/runnable/linkdebug.sh
+++ b/test/runnable/linkdebug.sh
@@ -1,0 +1,21 @@
+#!/usr/bin/env bash
+
+src=runnable${SEP}extra-files
+dir=${RESULTS_DIR}${SEP}runnable
+output_file=${dir}/linkdebug.sh.out
+
+if [ $OS == "win32" -o  $OS == "win64" ]; then
+	LIBEXT=.lib
+else
+	LIBEXT=.a
+fi
+libname=${dir}${SEP}libX${LIBEXT}
+
+$DMD -m${MODEL} -I${src} -of${libname} -lib ${src}${SEP}linkdebug_uni.d ${src}${SEP}linkdebug_range.d ${src}${SEP}linkdebug_primitives.d || exit 1
+
+$DMD -m${MODEL} -I${src} -of${dir}${SEP}linkdebug${EXE} -g -debug ${src}${SEP}linkdebug.d ${libname} || exit 1
+
+rm ${libname}
+rm ${dir}/{linkdebug${OBJ},linkdebug${EXE}}
+
+echo Success > ${output_file}

